core/trino-main/src/main/java/io/trino/sql/planner/ScopeAware.java

```diff
@@ -117,10 +116,9 @@ private Boolean scopeAwareComparison(Node left, Node right)
                 //    * left.hasOuterParent(...) == true and right.hasOuterParent(...) == false
                 //    * leftField.getFieldId().equals(rightField.getFieldId()) == true
                 // Both fields would seem to have different hashes but be equal to each other.
-                // However, this cannot happen because we *require* that both expressions being compared by
-                // rooted in the same "query scope" (i.e., sub-scopes that are local to each other) -- see ScopeAwareKey.equals().
-                // If both fields have the same field id, by definition they will produce the same result for hasOuterParent().
-                checkState(leftScope.hasOuterParent(queryScope) == rightScope.hasOuterParent(queryScope));
+                if (leftScope.hasOuterParent(queryScope) != rightScope.hasOuterParent(queryScope)) {
```

Comment:
Can we add some test cases to ensure this fix work ?

Reply:
Sure, but I'm not yet sure if this fix is actually suitable or there was something missing in https://github.com/trinodb/trino/pull/3600 which seems the root cause of the error. I'm now trying to understand what was changed in https://github.com/trinodb/trino/pull/3600.

Reply:
Added a test case.